usr/src/uts/common/brand/lx/syscall/lx_socket.c

```diff
@@ -240,7 +242,36 @@ static const int stol_socktype[SOCK_SEQPACKET + 1] = {
 #define	STOL_SOCKTYPE(t)	\
 	((t) <= SOCK_SEQPACKET ? stol_socktype[(t)] : SOCK_INVAL)
 
+/*
+ * illumos represents tcp states with ranges between -6 to 6, where:
+ * tcp states > 0 define connections that have been stablished.
+ * tcp states < 0 define connections that have not yet stablished.
+ * Where Linux defines their tcp states as a range from 1 to 12.
+ * the macro STOL_TCPSTATE translate our illumos tcp states to a Linux one,
+ * in case there is no tcp state returned by OS that matches Linux, the state
+ * LX_TCP_CLOSE will be returned.
+ */
+
+/* Table based on -6 to 6 being mapped on to 0 to 13 */
+static const int stol_tcp_state[LX_TCP_NEW_SYN_RECV + 1] = {
+	LX_TCP_CLOSE,		/* -6 ==> 0 */
+	LX_TCP_CLOSE,		/* -5 ==> 1, no TCPS_IDLE in Linux */
+	LX_TCP_LISTEN,		/* -4 ==> 2, no TCPS_BOUND in Linux */
```

Comment:
Did you experiment with Linux to obtain LX_TCP_LISTEN state on a bound but not listening socket?

Reply:
I have a sample C program to test some scenarios: 
the results match in TCP states between native Linux and an LX branded zone, but the values for other struct members for tcp_info in this patch need more work in converting the values to Linux expected metrics.

LX Branded Zone
```bash
test_connect[SUCCESS] TCP state: LX_TCP_ESTABLISHED
test_connect[SUCCESS] tcpi_rto  1000
test_connect[SUCCESS] tcpi_last_data_recv 13125255
test_connect[SUCCESS] tcpi_rtt  4000000
test_connect[SUCCESS] tcpi_rttvar 500000
test_connect[SUCCESS] tcpi_snd_ssthresh 32975
test_connect[SUCCESS] tcpi_snd_cwnd  16360
test_connect[SUCCESS] tcpi_snd_mss  8180
test_connect[SUCCESS] tcpi_unacked 1289948609
test_connect[SUCCESS] tcpi_sacked 0
test_connect[SUCCESS] tcpi_pmtu  8252
test_connect[SUCCESS] tcpi_total_retrans 1289948609
test_connect[SUCCESS] tcpi_unacked  1289948609
test_connect[SUCCESS] tcpi_sacked  0
test_connect[SUCCESS] tcpi_rcv_space 5
test_connect[SUCCESS]:SOCK_STREAM NOT_CONNECTED TCP state: LX_TCP_CLOSE
test_connect[SUCCESS] tcpi_rto  1000
test_connect[SUCCESS] tcpi_last_data_recv 0
test_connect[SUCCESS] tcpi_rtt  4000000
test_connect[SUCCESS] tcpi_rttvar 500000
test_connect[SUCCESS] tcpi_snd_ssthresh 0
test_connect[SUCCESS] tcpi_snd_cwnd  0
test_connect[SUCCESS] tcpi_snd_mss  536
test_connect[SUCCESS] tcpi_unacked 0
test_connect[SUCCESS] tcpi_sacked 0
test_connect[SUCCESS] tcpi_pmtu  0
test_connect[SUCCESS] tcpi_total_retrans 0
test_connect[SUCCESS] tcpi_unacked  0
test_connect[SUCCESS] tcpi_sacked  0
test_connect[SUCCESS] tcpi_rcv_space 14
test_connect[FAIL] SOCK_STREAM CONNECTED: connect() failed: Address family not supported by protocol
test_connect[SUCCESS]:SOCK_STREAM CONNECTED TCP state: LX_TCP_ESTABLISHED
test_connect[SUCCESS] tcpi_rto  1000
test_connect[SUCCESS] tcpi_last_data_recv 13125256
test_connect[SUCCESS] tcpi_rtt  4000000
test_connect[SUCCESS] tcpi_rttvar 500000
test_connect[SUCCESS] tcpi_snd_ssthresh 32975
test_connect[SUCCESS] tcpi_snd_cwnd  16360
test_connect[SUCCESS] tcpi_snd_mss  8180
test_connect[SUCCESS] tcpi_unacked -253238741
test_connect[SUCCESS] tcpi_sacked 0
test_connect[SUCCESS] tcpi_pmtu  8232
test_connect[SUCCESS] tcpi_total_retrans -253238741
test_connect[SUCCESS] tcpi_unacked  -253238741
test_connect[SUCCESS] tcpi_sacked  0
test_connect[SUCCESS] tcpi_rcv_space 5
test_bound[SUCCESS] BOUND tcp state: LX_TCP_CLOSE
test_bound[SUCCESS] tcpi_rto  1000
test_bound[SUCCESS] tcpi_last_data_recv 0
test_bound[SUCCESS] tcpi_rtt  4000000
test_bound[SUCCESS] tcpi_rttvar 500000
test_bound[SUCCESS] tcpi_snd_ssthresh 0
test_bound[SUCCESS] tcpi_snd_cwnd  0
test_bound[SUCCESS] tcpi_snd_mss  536
test_bound[SUCCESS] tcpi_unacked 0
test_bound[SUCCESS] tcpi_sacked 0
test_bound[SUCCESS] tcpi_pmtu  0
test_bound[SUCCESS] tcpi_total_retrans 0
test_bound[SUCCESS] tcpi_unacked  0
test_bound[SUCCESS] tcpi_sacked  0
test_bound[SUCCESS] tcpi_rcv_space 14
test_bound[SUCCESS] LISTEN tcp state: LX_TCP_LISTEN
test_bound[SUCCESS] tcpi_rto  1000
test_bound[SUCCESS] tcpi_last_data_recv 0
test_bound[SUCCESS] tcpi_rtt  4000000
test_bound[SUCCESS] tcpi_rttvar 500000
test_bound[SUCCESS] tcpi_snd_ssthresh 0
test_bound[SUCCESS] tcpi_snd_cwnd  0
test_bound[SUCCESS] tcpi_snd_mss  536
test_bound[SUCCESS] tcpi_unacked 0
test_bound[SUCCESS] tcpi_sacked 0
test_bound[SUCCESS] tcpi_pmtu  0
test_bound[SUCCESS] tcpi_total_retrans 0
test_bound[SUCCESS] tcpi_unacked  0
test_bound[SUCCESS] tcpi_sacked  0
test_bound[SUCCESS] tcpi_rcv_space 14
``` 

Native Linux 

```bash
test_connect[SUCCESS] TCP state: LX_TCP_ESTABLISHED
test_connect[SUCCESS] tcpi_rto  200000
test_connect[SUCCESS] tcpi_last_data_recv 0
test_connect[SUCCESS] tcpi_rtt  17
test_connect[SUCCESS] tcpi_rttvar 8
test_connect[SUCCESS] tcpi_snd_ssthresh 2147483647
test_connect[SUCCESS] tcpi_snd_cwnd  10
test_connect[SUCCESS] tcpi_snd_mss  32732
test_connect[SUCCESS] tcpi_unacked 0
test_connect[SUCCESS] tcpi_sacked 0
test_connect[SUCCESS] tcpi_pmtu  65536
test_connect[SUCCESS] tcpi_total_retrans 0
test_connect[SUCCESS] tcpi_unacked  0
test_connect[SUCCESS] tcpi_sacked  0
test_connect[SUCCESS] tcpi_rcv_space 65476
test_connect[SUCCESS]:SOCK_STREAM NOT_CONNECTED TCP state: LX_TCP_CLOSE
test_connect[SUCCESS] tcpi_rto  1000000
test_connect[SUCCESS] tcpi_last_data_recv 366854276
test_connect[SUCCESS] tcpi_rtt  0
test_connect[SUCCESS] tcpi_rttvar 250000
test_connect[SUCCESS] tcpi_snd_ssthresh 2147483647
test_connect[SUCCESS] tcpi_snd_cwnd  10
test_connect[SUCCESS] tcpi_snd_mss  536
test_connect[SUCCESS] tcpi_unacked 0
test_connect[SUCCESS] tcpi_sacked 0
test_connect[SUCCESS] tcpi_pmtu  0
test_connect[SUCCESS] tcpi_total_retrans 0
test_connect[SUCCESS] tcpi_unacked  0
test_connect[SUCCESS] tcpi_sacked  0
test_connect[SUCCESS] tcpi_rcv_space 0
test_connect[FAIL] SOCK_STREAM CONNECTED: connect() failed: Address family not supported by protocol
test_connect[SUCCESS]:SOCK_STREAM CONNECTED TCP state: LX_TCP_ESTABLISHED
test_connect[SUCCESS] tcpi_rto  200000
test_connect[SUCCESS] tcpi_last_data_recv 0
test_connect[SUCCESS] tcpi_rtt  12
test_connect[SUCCESS] tcpi_rttvar 6
test_connect[SUCCESS] tcpi_snd_ssthresh 2147483647
test_connect[SUCCESS] tcpi_snd_cwnd  10
test_connect[SUCCESS] tcpi_snd_mss  32741
test_connect[SUCCESS] tcpi_unacked 0
test_connect[SUCCESS] tcpi_sacked 0
test_connect[SUCCESS] tcpi_pmtu  65535
test_connect[SUCCESS] tcpi_total_retrans 0
test_connect[SUCCESS] tcpi_unacked  0
test_connect[SUCCESS] tcpi_sacked  0
test_connect[SUCCESS] tcpi_rcv_space 65495
test_bound[SUCCESS] BOUND tcp state: LX_TCP_CLOSE
test_bound[SUCCESS] tcpi_rto  1000000
test_bound[SUCCESS] tcpi_last_data_recv 366854276
test_bound[SUCCESS] tcpi_rtt  0
test_bound[SUCCESS] tcpi_rttvar 250000
test_bound[SUCCESS] tcpi_snd_ssthresh 2147483647
test_bound[SUCCESS] tcpi_snd_cwnd  10
test_bound[SUCCESS] tcpi_snd_mss  536
test_bound[SUCCESS] tcpi_unacked 0
test_bound[SUCCESS] tcpi_sacked 0
test_bound[SUCCESS] tcpi_pmtu  0
test_bound[SUCCESS] tcpi_total_retrans 0
test_bound[SUCCESS] tcpi_unacked  0
test_bound[SUCCESS] tcpi_sacked  0
test_bound[SUCCESS] tcpi_rcv_space 0
test_bound[SUCCESS] LISTEN tcp state: LX_TCP_LISTEN
test_bound[SUCCESS] tcpi_rto  0
test_bound[SUCCESS] tcpi_last_data_recv 0
test_bound[SUCCESS] tcpi_rtt  0
test_bound[SUCCESS] tcpi_rttvar 0
test_bound[SUCCESS] tcpi_snd_ssthresh 0
test_bound[SUCCESS] tcpi_snd_cwnd  10
test_bound[SUCCESS] tcpi_snd_mss  0
test_bound[SUCCESS] tcpi_unacked 0
test_bound[SUCCESS] tcpi_sacked 10
test_bound[SUCCESS] tcpi_pmtu  0
test_bound[SUCCESS] tcpi_total_retrans 0
test_bound[SUCCESS] tcpi_unacked  0
test_bound[SUCCESS] tcpi_sacked  10
test_bound[SUCCESS] tcpi_rcv_space 0
```


```c
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <netdb.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>




int test_connect (int argc, char *argv[], int test);
int test_bound (int argc, char *argv[], int test);
char *state2string (int state);

int
main (int argc, char *argv[])
{
  test_connect (argc, argv, 1);
  test_connect (argc, argv, 2);
  test_connect (argc, argv, 3);
  test_bound (argc, argv, 1);
  test_bound (argc, argv, 2);

}


int
test_bound (int argc, char *argv[], int test)
{

  int listenfd = 0, connfd = 0;
  struct sockaddr_in serv_addr;

  listenfd = socket (AF_INET, SOCK_STREAM, 0);
  memset (&serv_addr, '0', sizeof (serv_addr));

  serv_addr.sin_family = AF_INET;
  serv_addr.sin_addr.s_addr = htonl (INADDR_ANY);
  serv_addr.sin_port = htons (5000);

  bind (listenfd, (struct sockaddr *) &serv_addr, sizeof (serv_addr));
  char name[40] = "";
  int namelen = sizeof (struct tcp_info);
  struct tcp_info tcpi;

  switch (test)
    {

    case 1:
      if (getsockopt (listenfd, IPPROTO_TCP, TCP_INFO, &tcpi, &namelen) == -1)
	{
	  perror ("test_bound[FAIL] BOUND getsockopt(TCP_INFO) failed");
	  return (1);
	}
      printf ("test_bound[SUCCESS] BOUND tcp state: %s\n",
	      state2string (tcpi.tcpi_state));
      break;
    case 2:
      listen (listenfd, 10);

      if (getsockopt (listenfd, IPPROTO_TCP, TCP_INFO, &tcpi, &namelen) == -1)
	{
	  perror ("test_bound[FAIL] LISTEN: getsockopt(TCP_INFO) failed");
	  return (1);
	}
      printf ("test_bound[SUCCESS] LISTEN tcp state: %s\n",
	      state2string (tcpi.tcpi_state));
      break;
    }

  printf ("test_bound[SUCCESS] tcpi_rto  %d\n", tcpi.tcpi_rto);
  printf ("test_bound[SUCCESS] tcpi_last_data_recv %d\n",
	  tcpi.tcpi_last_data_recv);
  printf ("test_bound[SUCCESS] tcpi_rtt  %d\n", tcpi.tcpi_rtt);
  printf ("test_bound[SUCCESS] tcpi_rttvar %d\n", tcpi.tcpi_rttvar);
  printf ("test_bound[SUCCESS] tcpi_snd_ssthresh %d\n",
	  tcpi.tcpi_snd_ssthresh);
  printf ("test_bound[SUCCESS] tcpi_snd_cwnd  %d\n", tcpi.tcpi_snd_cwnd);
  printf ("test_bound[SUCCESS] tcpi_snd_mss  %d\n", tcpi.tcpi_snd_mss);
  printf ("test_bound[SUCCESS] tcpi_unacked %d\n", tcpi.tcpi_unacked);
  printf ("test_bound[SUCCESS] tcpi_sacked %d\n", tcpi.tcpi_sacked);
  printf ("test_bound[SUCCESS] tcpi_pmtu  %d\n", tcpi.tcpi_pmtu);
  printf ("test_bound[SUCCESS] tcpi_total_retrans %d\n",
	  tcpi.tcpi_total_retrans);
  printf ("test_bound[SUCCESS] tcpi_unacked  %d\n", tcpi.tcpi_unacked);
  printf ("test_bound[SUCCESS] tcpi_sacked  %d\n", tcpi.tcpi_sacked);
  printf ("test_bound[SUCCESS] tcpi_rcv_space %d\n", tcpi.tcpi_rcv_space);

}

int
test_connect (int argc, char *argv[], int test)
{
  struct addrinfo hints, *gair, *p;
  int fd, rv, rlen;

  if (argc != 3)
    {
      fprintf (stderr, "%s <host> <port>\n", argv[0]);
      return (1);
    }

  memset (&hints, 0, sizeof (hints));
  hints.ai_family = PF_UNSPEC;
  hints.ai_socktype = SOCK_STREAM;

  if ((rv = getaddrinfo (argv[1], argv[2], &hints, &gair)) != 0)
    {
      fprintf (stderr, "test_connect[FAIL]: getaddrinfo() failed: %s\n",
	       gai_strerror (rv));
      return (1);
    }

  for (p = gair; p != NULL; p = p->ai_next)
    {
      switch (test)
	{
	case 1:
	  if ((fd =
	       socket (p->ai_family, p->ai_socktype, p->ai_protocol)) == -1)
	    {
	      perror ("test_connect[FAIL]: socket() failed");
	      continue;
	    }
	  if (connect (fd, p->ai_addr, p->ai_addrlen) == -1)
	    {
	      close (fd);
	      perror ("test_connect[FAIL]: connect() failed");
	      continue;
	    }
	  break;

	case 2:
	  if ((fd = socket (AF_INET, SOCK_STREAM, 0)) == -1)
	    {
	      perror
		("test_connect[FAIL] SOCK_STREAM NOT_CONNECTED: socket() failed");
	      continue;
	    }
	  break;

	case 3:
	  if ((fd = socket (AF_INET, SOCK_STREAM, 0)) == -1)
	    {
	      perror
		("test_connect[FAIL] SOCK_STREAM CONNECTED: socket() failed");
	      continue;
	    }
	  if (connect (fd, p->ai_addr, p->ai_addrlen) == -1)
	    {
	      close (fd);
	      perror
		("test_connect[FAIL] SOCK_STREAM CONNECTED: connect() failed");
	      continue;
	    }
	  break;
	}
      break;
    }

  if (p == NULL)
    {
      fprintf (stderr, "test_connect[FAIL]:  failed to connect to server\n");
      return (1);
    }

  freeaddrinfo (gair);

  int namelen = sizeof (struct tcp_info);
  struct tcp_info tcpi;
  if (getsockopt (fd, IPPROTO_TCP, TCP_INFO, &tcpi, &namelen) == -1)
    {
      switch (test)
	{
	case 1:
	  perror ("test_connect[FAIL]:getsockopt(TCP_INFO) failed");
	  break;

	case 2:
	  perror
	    ("test_connect[FAIL] SOCK_STREAM NOT_CONNECTED :getsockopt(TCP_INFO) failed");
	  break;

	case 3:
	  perror
	    ("test_connect[FAIL] SOCK_STREAM CONNECTED :getsockopt(TCP_INFO) failed");
	  break;

	default:
	  perror
	    ("test_connect[FAIL] SOCK_STREAM :getsockopt(TCP_INFO) failed");
	}
      return -1;
    }

  switch (test)
    {
    default:
    case 1:
      printf ("test_connect[SUCCESS] TCP state: %s\n",
	      state2string (tcpi.tcpi_state));
      break;

    case 2:
      printf
	("test_connect[SUCCESS]:SOCK_STREAM NOT_CONNECTED TCP state: %s\n",
	 state2string (tcpi.tcpi_state));
      break;

    case 3:
      printf
	("test_connect[SUCCESS]:SOCK_STREAM CONNECTED TCP state: %s\n",
	 state2string (tcpi.tcpi_state));
    }

  printf ("test_connect[SUCCESS] tcpi_rto  %d\n", tcpi.tcpi_rto);
  printf ("test_connect[SUCCESS] tcpi_last_data_recv %d\n",
	  tcpi.tcpi_last_data_recv);
  printf ("test_connect[SUCCESS] tcpi_rtt  %d\n", tcpi.tcpi_rtt);
  printf ("test_connect[SUCCESS] tcpi_rttvar %d\n", tcpi.tcpi_rttvar);
  printf ("test_connect[SUCCESS] tcpi_snd_ssthresh %d\n",
	  tcpi.tcpi_snd_ssthresh);
  printf ("test_connect[SUCCESS] tcpi_snd_cwnd  %d\n", tcpi.tcpi_snd_cwnd);
  printf ("test_connect[SUCCESS] tcpi_snd_mss  %d\n", tcpi.tcpi_snd_mss);
  printf ("test_connect[SUCCESS] tcpi_unacked %d\n", tcpi.tcpi_unacked);
  printf ("test_connect[SUCCESS] tcpi_sacked %d\n", tcpi.tcpi_sacked);
  printf ("test_connect[SUCCESS] tcpi_pmtu  %d\n", tcpi.tcpi_pmtu);
  printf ("test_connect[SUCCESS] tcpi_total_retrans %d\n",
	  tcpi.tcpi_total_retrans);
  printf ("test_connect[SUCCESS] tcpi_unacked  %d\n", tcpi.tcpi_unacked);
  printf ("test_connect[SUCCESS] tcpi_sacked  %d\n", tcpi.tcpi_sacked);
  printf ("test_connect[SUCCESS] tcpi_rcv_space %d\n", tcpi.tcpi_rcv_space);
}


/* TCP states */
#define	TCPS_CLOSED		-6
#define	TCPS_IDLE		-5	/* idle (opened, but not bound) */
#define	TCPS_BOUND		-4	/* bound, ready to connect or accept */
#define	TCPS_LISTEN		-3	/* listening for connection */
#define	TCPS_SYN_SENT		-2	/* active, have sent syn */
#define	TCPS_SYN_RCVD		-1	/* have received syn (and sent ours) */
/* states < TCPS_ESTABLISHED are those where connections not established */
#define	TCPS_ESTABLISHED	0	/* established */
#define	TCPS_CLOSE_WAIT		1	/* rcvd fin, waiting for close */
/* states > TCPS_CLOSE_WAIT are those where user has closed */
#define	TCPS_FIN_WAIT_1		2	/* have closed and sent fin */
#define	TCPS_CLOSING		3	/* closed, xchd FIN, await FIN ACK */
#define	TCPS_LAST_ACK		4	/* had fin and close; await FIN ACK */
/* states > TCPS_CLOSE_WAIT && < TCPS_FIN_WAIT_2 await ACK of FIN */
#define	TCPS_FIN_WAIT_2		5	/* have closed, fin is acked */
#define	TCPS_TIME_WAIT		6	/* in 2*msl quiet wait after close */



#define LX_TCP_ESTABLISHED  1
#define LX_TCP_SYN_SENT  2
#define LX_TCP_SYN_RECV  3
#define LX_TCP_FIN_WAIT1  4
#define LX_TCP_FIN_WAIT2  5
#define LX_TCP_TIME_WAIT  6
#define LX_TCP_CLOSE  7
#define LX_TCP_CLOSE_WAIT  8
#define LX_TCP_LAST_ACK  9
#define LX_TCP_LISTEN  10
#define LX_TCP_CLOSING  11
#define LX_TCP_NEW_SYN_RECV  12

char *
state2string (int state)
{
  static char st[32];
  switch (state)
    {
    case LX_TCP_ESTABLISHED:
      sprintf (st, "LX_TCP_ESTABLISHED");
      break;
    case LX_TCP_SYN_SENT:
      sprintf (st, "LX_TCP_SYN_SENT");
      break;
    case LX_TCP_SYN_RECV:
      sprintf (st, "LX_TCP_SYN_RECV");
      break;
    case LX_TCP_FIN_WAIT1:
      sprintf (st, "LX_TCP_FIN_WAIT1");
      break;
    case LX_TCP_FIN_WAIT2:
      sprintf (st, "LX_TCP_FIN_WAIT2");
      break;
    case LX_TCP_TIME_WAIT:
      sprintf (st, "LX_TCP_TIME_WAIT");
      break;
    case LX_TCP_CLOSE:
      sprintf (st, "LX_TCP_CLOSE");
      break;

    case LX_TCP_CLOSE_WAIT:
      sprintf (st, "LX_TCP_CLOSE_WAIT");
      break;

    case LX_TCP_LAST_ACK:
      sprintf (st, "LX_TCP_LAST_ACK");
      break;

    case LX_TCP_LISTEN:
      sprintf (st, "LX_TCP_LISTEN");
      break;

    case LX_TCP_CLOSING:
      sprintf (st, "LX_TCP_CLOSING");
      break;
    case LX_TCP_NEW_SYN_RECV:
      sprintf (st, "LX_TCP_NEW_SYN_RECV");
      break;
    }
  return st;
}
```